f3.go

```diff
@@ -1,141 +1,70 @@
 	"github.com/filecoin-project/go-f3/gpbft"
 	"github.com/filecoin-project/go-f3/manifest"
-	"github.com/ipfs/go-datastore"
-	"github.com/ipfs/go-datastore/namespace"
 
-	logging "github.com/ipfs/go-log/v2"
 	pubsub "github.com/libp2p/go-libp2p-pubsub"
 	"github.com/libp2p/go-libp2p/core/host"
-	peer "github.com/libp2p/go-libp2p/core/peer"
 
+	"github.com/Kubuxu/go-broadcast"
+	"github.com/ipfs/go-datastore"
 	"go.uber.org/multierr"
+	"golang.org/x/sync/errgroup"
 	"golang.org/x/xerrors"
 )
 
 type F3 struct {
-	// mLk protects the runner and certStore
-	mLk    sync.Mutex
-	runner *gpbftRunner
-	// certStore is nil until Run is called on the F3
-	csLk      sync.Mutex
-	certStore *certstore.Store
-
-	cancelCtx context.CancelFunc
-	host      host.Host
-	ds        datastore.Datastore
-	ec        ec.Backend
-	log       Logger
-
-	// psLk protects pubsub. We need an independent one
-	// due to its use in Broadcast
-	psLk   sync.Mutex
-	pubsub *pubsub.PubSub
-	msgSub *pubsub.Subscription
-
-	client *client
-}
-
-type client struct {
-	// certStore is nil until Run is called on the F3
-	certStore *certstore.Store
-	manifest  manifest.ManifestProvider
-	ec        ec.Backend
-
+	verifier          gpbft.Verifier
 	signingMarshaller gpbft.SigningMarshaler
+	manifestProvider  manifest.ManifestProvider
 
 	busBroadcast broadcast.Channel[*gpbft.MessageBuilder]
 
-	gpbft.Verifier
-	logger         Logger
-	loggerWithSkip Logger
-
-	// Populated after Run is called
-	messageQueue <-chan gpbft.ValidatedMessage
-	topic        *pubsub.Topic
-
-	// Notifies manifest updates
-	manifestUpdate <-chan uint64
-	// Triggers the cancellation of the incoming message
-	// routine to avoid delivering any outstanding messages.
-	incomingCancel func()
-}
-
-func (mc *client) BroadcastMessage(ctx context.Context, mb *gpbft.MessageBuilder) error {
-	mb.SetNetworkName(mc.manifest.Manifest().NetworkName)
-	mb.SetSigningMarshaler(mc.signingMarshaller)
-	mc.busBroadcast.Publish(mb)
-	return nil
-}
-
-func (mc *client) GetPowerTable(ctx context.Context, ts gpbft.TipSetKey) (gpbft.PowerEntries, error) {
-	// Apply power table deltas from the manifest
-	pt, err := mc.ec.GetPowerTable(ctx, ts)
-	if err != nil {
-		return nil, xerrors.Errorf("getting power table: %w", err)
-	}
-	return certs.ApplyPowerTableDiffs(pt, mc.manifest.Manifest().PowerUpdate)
-}
-
-func (mc *client) IncomingMessages() <-chan gpbft.ValidatedMessage {
-	return mc.messageQueue
-}
-
-var _ gpbft.Tracer = (*client)(nil)
+	host   host.Host
+	ds     datastore.Datastore
+	ec     ec.Backend
+	log    Logger
+	pubsub *pubsub.PubSub
 
-// Log fulfills the gpbft.Tracer interface
-func (mc *client) Log(fmt string, args ...any) {
-	mc.loggerWithSkip.Debugf(fmt, args...)
-}
+	runningCtx context.Context
+	cancelCtx  context.CancelFunc
+	errgrp     *errgroup.Group
 
-func (mc *client) Logger() Logger {
-	return mc.logger
+	mu       sync.Mutex
+	cs       *certstore.Store
+	manifest *manifest.Manifest
+	runner   *gpbftRunner
```

Comment:
I store these separately because:

1. The certificate-store remains even when the runner is stopped (paused).
2. The manifest can is set before bootstrap (if available).
3. The runner is unset whenever we pause.